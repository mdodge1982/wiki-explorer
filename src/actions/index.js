import socket from './websocket';

const listentoSocket = () => {
	return (dispatch, getState) => {
		socket.addEventListener('message', (e) => {
			//Convert socket message to redux event
			const message = JSON.parse(e.data);
			dispatch({type: message.name.toUpperCase(), id: message.id, data: message.data});
		});
	};
};

const socketRequest = (requestData) => {
	//Is the socket ready?
	const p = new Promise((resolve, reject) => {
		if (socket.readyState === 1) {
			resolve();
		} else {
			socket.addEventListener('open', resolve);
		}
	}).then(() => {
		socket.send(JSON.stringify(requestData));
	});
	return p;
};

socket.onopen = () => {
	//Keep the socket open by pinging
	setInterval(() => {
		socketRequest({id: 'ping', name: 'ping'});
	}, 20000);
};

const requestSocket = (requestData) => {
	return (dispatch, getState) => {
		//Notify that a request has been started
		dispatch({
			type: 'REQUEST_' + requestData.name.toUpperCase(),
			id: requestData.id
		});
		return socketRequest(requestData);
	};
};

const fetchProjects = () => {
	return requestSocket({id: 'allpages', name: 'project.list'});
};

const selectPages = (projectName) => {
	return requestSocket({
		id: projectName,
		name: 'page.list',
		args: {
			project: projectName
		}
	});
};

const toggleProjectSubscription = (projectName, subscribe) => {
	const suffix = subscribe
		? 'subscribe'
		: 'unsubscribe';
	return requestSocket({
		id: projectName,
		name: 'project.' + suffix,
		args: {
			project: projectName
		}
	});
};

const togglePageSubscription = (pageId, subscribe) => {
	const suffix = subscribe
		? 'subscribe'
		: 'unsubscribe';
	return requestSocket({
		id: pageId,
		name: 'page.' + suffix,
		args: {
			pageId
		}
	});
};

const selectProject = (projectName) => {
	return (dispatch, getState) => {
		const {byName} = getState();
		const project = byName[projectName];
		let type = 'SELECT_PROJECT';
		if (project.selected) {
			type = 'DESELECT_PROJECT'
		} else {
			//Fetch the pages if there are none yet
			if (project.pages.length === 0) {
				dispatch(selectPages(projectName));
			}
			//subscribe to the newly selected project
			dispatch(toggleProjectSubscription(projectName, true));
		}
		if (byName.selected) {
			//unsubscribe from the old selected project
			const oldProject = byName[byName.selected];
			dispatch(toggleProjectSubscription(oldProject.name, false));
		}
		dispatch({type, id: projectName});
	};
}

const selectPage = (page) => {
	return (dispatch, getState) => {
		dispatch({type: 'PAGE.UPDATE', data: page, isFetching: true});
		//Removing page.query for now since we already get the same info from pages.list
		//Keep the subscription in case of updates
		// dispatch(requestSocket({
		// 	id: page.pageid,
		// 	name: 'page.query',
		// 	args: {
		// 		pageId: page.pageid
		// 	}
		// }));
		dispatch(togglePageSubscription(page.pageid, true));
	}
}

const clearSelectedPage = () => {
	return (dispatch, getState) => {
		const {selectedPage} = getState();
		dispatch(togglePageSubscription(selectedPage.pageid, false));
		dispatch({type: 'CLEAR_SELECTED_PAGE'});
	};

}

export {listentoSocket, fetchProjects, selectProject, selectPage, clearSelectedPage};
