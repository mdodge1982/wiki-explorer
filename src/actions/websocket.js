const socket = new WebSocket("wss://wiki-meta-explorer.herokuapp.com");

export default socket;
