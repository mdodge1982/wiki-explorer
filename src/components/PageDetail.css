.PageDetail {
	position: fixed;
	left: 0;
	top: 0;
	width: 100%;
	height: 100%;
	background: rgba(0,0,0,.5);
}
.PageDetail .inner {
	position: fixed;
	top: 5em;
	left: 50%;
	margin-left: -20em;
	width: 40em;
	background: #fff;
	box-shadow: 0 1em 3em rgba(0,0,0,.5);
	padding: 1em 2em 3em;
}
.PageDetail button {
	border: none;
	background: transparent;
	font-size: 2em;
	position: absolute;
	top: -1.5em;
	right: 0;
	color: #fff;
}
dl {
	margin: 0 0 1em;
	padding: 0 1em 0 0;
	width: 32%;
}
dt {
	font-weight: bold;
}
dd {
	margin: 0;
	padding: 0;
	word-break: break-all;
}
.comment {
	flex-grow: 1;
}
.revision {
	clear: left;
	padding: 1em 1em 0;
	background: #f5f5f5;
}
