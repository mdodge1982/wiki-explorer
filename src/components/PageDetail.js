import React, {Component} from 'react';
import {connect} from 'react-redux';
import {clearSelectedPage} from '../actions';
import './PageDetail.css';

class PageDetailComonent extends Component {
	render() {
		const page = this.props.page;
		if (!page.pageid) {
			return '';
		}
		const revision = page.revisions[0];
		return (
			<div className="PageDetail">
				<div className="inner">
					<button onClick={() => this.close()}>&times;</button>
					<h1>{page.title}</h1>
					<div className="flex">
						<dl>
							<dt>
								Page ID
							</dt>
							<dd>
								{page.pageid}
							</dd>
						</dl>

						<dl>
							<dt>
								Language
							</dt>
							<dd>
								{page.pagelanguage}
								({page.pagelanguagedir})
							</dd>
						</dl>

						<dl>
							<dt>
								Length
							</dt>
							<dd>
								{page.length}
							</dd>
						</dl>
					</div>
					<div className="revision">
						<h3>Last Revision</h3>
						<div className="flex">
							<dl>
								<dt>
									Revision ID
								</dt>
								<dd>
									{page.lastrevid}
								</dd>
							</dl>

							<dl>
								<dt>
									Parent ID
								</dt>
								<dd>
									{revision.parentid}
								</dd>
							</dl>

							<dl>
								<dt>
									User
								</dt>
								<dd>
									{revision.user}
								</dd>
							</dl>

							<dl>
								<dt>
									Timestamp
								</dt>
								<dd>
									{revision.timestamp}
								</dd>
							</dl>

							<dl className="comment">
								<dt>
									Comment
								</dt>
								<dd>
									{revision.comment}
								</dd>
							</dl>
						</div>
					</div>
				</div>
			</div>
		);
	}
	close() {
		this.props.dispatch(clearSelectedPage());
	}
}

const PageDetail = connect(({selectedPage}) => ({page: selectedPage}))(PageDetailComonent);

export default PageDetail;
