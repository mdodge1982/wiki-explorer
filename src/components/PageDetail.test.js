import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import PageDetail from './PageDetail';
import {Provider} from 'react-redux';
import store from '../store';
import {selectPage} from '../actions';
import renderer from 'react-test-renderer';

jest.mock('../actions/websocket');

const page = {
	"pageid": 9020,
	"title": "Daisy Duck",
	"pagelanguage": "en",
	"pagelanguagedir": "ltr",
	"length": 37954,
	"lastrevid": 797710133,
	"revisions": [
		{
			"revid": 797710133,
			"parentid": 797709597,
			"user": "XXX.YYY.ZZZ.WWW",
			"timestamp": "2017-08-28T19:03:35Z",
			"comment": "Test Comment"
		}
	]
};

test('Test render', () => {
	store.dispatch(selectPage(page));
	const component = renderer.create(
		<Provider store={store}>
			<PageDetail/>
		</Provider>
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();

});

test('Button closes modal', () => {
	expect(store.getState().selectedPage.pageid).toEqual(9020);
	const component = ReactTestUtils.renderIntoDocument(
		<Provider store={store}>
			<PageDetail/>
		</Provider>
	);
	const button = ReactTestUtils.findRenderedDOMComponentWithTag(component, 'button');
	ReactTestUtils.Simulate.click(button);
	expect(store.getState().selectedPage.pageid).toBeUndefined();
});
