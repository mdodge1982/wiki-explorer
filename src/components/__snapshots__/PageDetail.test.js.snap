// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render 1`] = `
<div
  className="PageDetail"
>
  <div
    className="inner"
  >
    <button
      onClick={[Function]}
    >
      Ã—
    </button>
    <h1>
      Daisy Duck
    </h1>
    <div
      className="flex"
    >
      <dl>
        <dt>
          Page ID
        </dt>
        <dd>
          9020
        </dd>
      </dl>
      <dl>
        <dt>
          Language
        </dt>
        <dd>
          en
           (
          ltr
          )
        </dd>
      </dl>
      <dl>
        <dt>
          Length
        </dt>
        <dd>
          37954
        </dd>
      </dl>
    </div>
    <div
      className="revision"
    >
      <h3>
        Last Revision
      </h3>
      <div
        className="flex"
      >
        <dl>
          <dt>
             Revision ID
          </dt>
          <dd>
            797710133
          </dd>
        </dl>
        <dl>
          <dt>
             Parent ID
          </dt>
          <dd>
            797709597
          </dd>
        </dl>
        <dl>
          <dt>
             User
          </dt>
          <dd>
            XXX.YYY.ZZZ.WWW
          </dd>
        </dl>
        <dl>
          <dt>
             Timestamp
          </dt>
          <dd>
            2017-08-28T19:03:35Z
          </dd>
        </dl>
        <dl
          className="comment"
        >
          <dt>
             Comment
          </dt>
          <dd>
            Test Comment
          </dd>
        </dl>
      </div>
    </div>
  </div>
</div>
`;
